feat(ui): Add stackdriver links to logs tab

deadlycoconuts opened the pull request:
## Context
Similar to how the Merlin logs tab displays links to stored logs on stackdriver (Google Cloud Logging) in addition to live logs, this PR adds a panel to the router logs tab to display links to stored logs on stackdriver for the following components:
- router
- enricher (Docker enrichers only)
- ensembler (Docker and pyfunc ensemblers only)
- ensembler building logs (pyfunc ensemblers only)

![Screenshot 2024-06-12 at 12 37 09](https://github.com/caraml-dev/turing/assets/36802364/00675a78-2754-4a09-b149-35afc67bba64)

deadlycoconuts commented on ui/src/utils/createStackdriverUrl.js:
```diff
@@ -0,0 +1,36 @@
+const stackdriverAPI = "https://console.cloud.google.com/logs/viewer";
```
This file is almost an exact copy of this file from Merlin: https://github.com/caraml-dev/merlin/blob/main/ui/src/utils/createStackdriverUrl.js

leonlnj replied:
can add these as a comment in the code? Ref:url....

deadlycoconuts replied:
Okay just added that!

deadlycoconuts commented on ui/src/components/pod_logs_viewer/PodLogsViewer.js:
```diff
@@ -110,33 +114,59 @@ export const PodLogsViewer = ({
+      {
+        Object.keys(stackdriverUrls).length !== 0 &&
+        (
+          <>
+            <EuiPanel>
+              <EuiFlexGroup direction="row" alignItems="center">
+                <EuiFlexItem style={{marginTop:0, marginBottom:0}} grow={false}>
+                  <EuiText  style={{ fontSize: '14px', fontWeight:"bold"}}>Stackdriver Logs</EuiText>
+                </EuiFlexItem>
+                {Object.entries(stackdriverUrls).map(([component,url])=> (
+                  <EuiFlexItem style={{marginTop:0, marginBottom:0, paddingLeft:"10px", textTransform: "capitalize"}} key={component} grow={false}>
+                    <EuiText size="xs" >
+                      <EuiLink href={url} target="_blank" external>{component.replace(new RegExp("_", "g"), " ")}</EuiLink>
+                    </EuiText>
+                  </EuiFlexItem>
+                ))}
+              </EuiFlexGroup>
+            </EuiPanel>
+            <EuiSpacer size="s" />
+          </>
```
This chunk is copied almost entirely as-in from Merlin here: https://github.com/caraml-dev/merlin/blob/main/ui/src/components/logs/ContainerLogsView.js#L193

deadlycoconuts commented on ui/src/router/details/logs/RouterLogsView.js:
```diff
@@ -55,24 +67,73 @@ export const RouterLogsView = ({ router }) => {
+          gcp_project: environment.gcp_project,
+          cluster: environment.cluster,
+          namespace: currentProject.name,
+          pod_name: router.name + "-turing-router-"  + router.config.version,
```
This naming convention deliberately ignores the knative revision number (`0` by default if deployed by the API server, anything else if the Knative spec is modified and saved manually):

Example of a pod name:
```test-router-1-turing-router-9-0-deployment-6f456745f7-xdr8x````

deadlycoconuts commented on ui/src/router/details/logs/RouterLogsView.js:
```diff
@@ -55,24 +67,73 @@ export const RouterLogsView = ({ router }) => {
+        if (router.config.ensembler.type === "pyfunc" && ensembler) {
+          urls["ensembler_image_builder"] = createStackdriverUrl({
+            job_name: "service-" + currentProject.name + "-" + ensembler.name,
+            start_time: ensembler.updated_at,
```
I set the `start_time` as the time after the `ensembler` has been updated because the `router.updated_at` time is always AFTER the image building jobs has been completed so there's no convenient timestamp I can use to set as the start time for the image building job.

leonlnj commented on ui/src/router/details/logs/RouterLogsView.js:
```diff
@@ -55,24 +67,73 @@ export const RouterLogsView = ({ router }) => {
+
+        // set image builder url
+        if (router.config.ensembler.type === "pyfunc" && ensembler) {
+          urls["ensembler_image_builder"] = createStackdriverUrl({
```
what about router image building?

leonlnj commented on ui/src/router/details/logs/RouterLogsView.js:
```diff
@@ -55,24 +67,73 @@ export const RouterLogsView = ({ router }) => {
+  useEffect(
+    () => {
+      let urls = {}
+      if (appConfig.imagebuilder.cluster && appConfig.imagebuilder.gcp_project && appConfig.imagebuilder.namespace &&
```
does this need to be here for can `createStackdriverUrl` import it?

deadlycoconuts replied:
It has to be there since Turing can only access the app configs via a config provider https://github.com/caraml-dev/turing/blob/91e22116cad1eb3b70ec100be94d683c8fa0b092/ui/src/config.js#L115 unlike Merlin which exposes the app configs as an exported variable: https://github.com/caraml-dev/merlin/blob/8edef22b29d0bfb2728d62b1f880f1f753f9509e/ui/src/config.js#L29. 

I can make the app configs exported also, but to be honest I'm not sure why we're using a provider if we're already exposing it globally 😅

deadlycoconuts replied:
Okay I've decided to just export the app configs, just like how it's done in Merlin. This keeps the code consistent across both codebases.

leonlnj reviewed (approved):
LGTM overall, thanks!

leonlnj commented on ui/src/router/details/logs/RouterLogsView.js:
```diff
@@ -55,24 +67,73 @@ export const RouterLogsView = ({ router }) => {
+  useEffect(
+    () => {
+      let urls = {}
+      if (appConfig.imagebuilder.cluster && appConfig.imagebuilder.gcp_project && appConfig.imagebuilder.namespace &&
```
nit: indent

deadlycoconuts replied:
Ok turned it to following to make it clearer:
```js
      if (
          appConfig.imagebuilder.cluster &&
          appConfig.imagebuilder.gcp_project &&
          appConfig.imagebuilder.namespace &&
          currentProject
      ) {
```

deadlycoconuts commented:
Okay I've just addressed all the comments; will merge this now. Sadly codecov is still failing with `62.26% (-5.83%) compared to 91e2211` even though there aren't any changes to the API 🥲
